File: backend/src/test/java/com/pos/core/services/DiscountPricingTest.java
package com.pos.core.services;

import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

import static org.assertj.core.api.Assertions.assertThat;

class DiscountPricingTest {

    @Test
    void priceLine_skipsGlobalWhenLineHasItemDiscount() {
        // 10% item on $100 → $90; global does not stack on item-discounted lines.
        DiscountPricing.PricedLine line = DiscountPricing.priceLine(
                new BigDecimal("100.0000"),
                new BigDecimal("1.0000"),
                new BigDecimal("0.1000"),
                new BigDecimal("0.1000"),
                false
        );

        assertThat(line.finalUnitPrice()).isEqualByComparingTo("90.0000");
        assertThat(line.lineDiscountAmount()).isEqualByComparingTo("10.0000");
    }

    @Test
    void priceLine_appliesGlobalOnlyWhenNoItemDiscountAndNotExcluded() {
        DiscountPricing.PricedLine eligible = DiscountPricing.priceLine(
                new BigDecimal("10.0000"),
                new BigDecimal("1.0000"),
                BigDecimal.ZERO,
                new BigDecimal("0.1000"),
                false
        );

        assertThat(eligible.finalUnitPrice()).isEqualByComparingTo("9.0000");
    }

    @Test
    void priceLine_skipsGlobalDiscountWhenProductIsExcluded() {
        DiscountPricing.PricedLine excluded = DiscountPricing.priceLine(
                new BigDecimal("10.0000"),
                new BigDecimal("1.0000"),
                BigDecimal.ZERO,
                new BigDecimal("0.1000"),
                true
        );

        assertThat(excluded.finalUnitPrice()).isEqualByComparingTo("10.0000");
    }

    @Test
    void priceLine_mixedCartItemDiscountAndGlobalOnDifferentLines() {
        // Cola: 10% item only (global skipped). Chips: 10% global only. Special: product excluded.
        DiscountPricing.PricedLine cola = DiscountPricing.priceLine(
                new BigDecimal("1.9900"),
                new BigDecimal("1.0000"),
                new BigDecimal("0.1000"),
                new BigDecimal("0.1000"),
                false
        );
        DiscountPricing.PricedLine chips = DiscountPricing.priceLine(
                new BigDecimal("2.5000"),
                new BigDecimal("1.0000"),
                BigDecimal.ZERO,
                new BigDecimal("0.1000"),
                false
        );
        DiscountPricing.PricedLine special = DiscountPricing.priceLine(
                new BigDecimal("2.5000"),
                new BigDecimal("1.0000"),
                BigDecimal.ZERO,
                new BigDecimal("0.1000"),
                true
        );

        assertThat(cola.finalUnitPrice()).isEqualByComparingTo("1.7910");
        assertThat(chips.finalUnitPrice()).isEqualByComparingTo("2.2500");
        assertThat(special.finalUnitPrice()).isEqualByComparingTo("2.5000");

        BigDecimal subtotal = cola.lineTotal().add(chips.lineTotal()).add(special.lineTotal());
        BigDecimal totalDiscount = cola.lineDiscountAmount()
                .add(chips.lineDiscountAmount())
                .add(special.lineDiscountAmount());

        assertThat(subtotal).isEqualByComparingTo("6.5410");
        assertThat(totalDiscount).isEqualByComparingTo("0.4490");
    }
}

File: frontend/src/lib/discountPricing.test.ts
import { describe, expect, it } from 'vitest'
import { priceLine } from '@/lib/discountPricing'

describe('discountPricing', () => {
  it('skips global when line has item discount', () => {
    const line = priceLine(100, 1, 0.1, 0.1, false)
    expect(line.finalUnitPrice).toBe(90)
    expect(line.lineDiscountAmount).toBe(10)
  })

  it('applies global only when no item discount and not excluded', () => {
    const eligible = priceLine(10, 1, 0, 0.1, false)
    const excluded = priceLine(10, 1, 0, 0.1, true)
    expect(eligible.finalUnitPrice).toBe(9)
    expect(excluded.finalUnitPrice).toBe(10)
  })

  it('mixed cart matches backend DiscountPricingTest', () => {
    const cola = priceLine(1.99, 1, 0.1, 0.1, false)
    const chips = priceLine(2.5, 1, 0, 0.1, false)
    const special = priceLine(2.5, 1, 0, 0.1, true)

    expect(cola.finalUnitPrice).toBe(1.791)
    expect(chips.finalUnitPrice).toBe(2.25)
    expect(special.finalUnitPrice).toBe(2.5)

    const subtotal = cola.lineTotal + chips.lineTotal + special.lineTotal
    const totalDiscount =
      cola.lineDiscountAmount + chips.lineDiscountAmount + special.lineDiscountAmount

    expect(subtotal).toBe(6.541)
    expect(totalDiscount).toBe(0.449)
  })
})
